builtin/logical/aws/path_roles.go

```diff
@@ -58,94 +85,342 @@ func pathRoles() *framework.Path {
+				PolicyArns:               []string{entry},
+				ProhibitFlexibleCredPath: true,
+				Version:                  1,
+			}
+		default:
+			newRoleEntry = &awsRoleEntry{
+				InvalidData: entry,
+				Version:     1,
+			}
+		}
 	} else {
-		// Write the arn ref into storage
-		err := req.Storage.Put(ctx, &logical.StorageEntry{
-			Key:   "policy/" + d.Get("name").(string),
-			Value: []byte(d.Get("arn").(string)),
-		})
+		compacted, err := compactJSON(entry)
 		if err != nil {
-			return nil, err
+			newRoleEntry = &awsRoleEntry{
+				InvalidData: entry,
+				Version:     1,
+			}
+		} else {
+			// unfortunately, this is ambiguous between the cred types, so allow both
+			newRoleEntry = &awsRoleEntry{
+				CredentialTypes:          []string{iamUserCred, federationTokenCred},
+				PolicyDocument:           compacted,
+				ProhibitFlexibleCredPath: true,
+				Version:                  1,
+			}
 		}
 	}
 
-	return nil, nil
+	return newRoleEntry
 }
 
+func nonLockedSetAwsRole(ctx context.Context, s logical.Storage, roleName string, roleEntry *awsRoleEntry) error {
+	if roleName == "" {
+		return fmt.Errorf("empty role name")
+	}
+	if roleEntry == nil {
+		return fmt.Errorf("nil roleEntry")
+	}
+	entry, err := logical.StorageEntryJSON("role/"+roleName, roleEntry)
+	if err != nil {
+		return err
+	}
+	if entry == nil {
+		return fmt.Errorf("nil result when writing to storage")
+	}
+	if err := s.Put(ctx, entry); err != nil {
+		return err
+	}
+	return nil
+}
+
+type awsRoleEntry struct {
+	CredentialTypes          []string `json:"credential_types"`                      // Entries must all be in the set of ("iam_user", "assumed_role", "federation_token")
+	PolicyArns               []string `json:"policy_arns"`                           // ARNs of managed policies to attach to an IAM user
+	RoleArns                 []string `json:"role_arns"`                             // ARNs of roles to assume for AssumedRole credentials
+	PolicyDocument           string   `json:"policy_document"`                       // JSON-serialized inline policy to attach to IAM users and/or to specify as the Policy parameter in AssumeRole calls
+	InvalidData              string   `json:"invalid_data,omitempty"`                // Invalid role data. Exists to support converting the legacy role data into the new format
+	ProhibitFlexibleCredPath bool     `json:"prohibit_flexible_cred_path,omitempty"` // Disallow accessing STS credentials via the creds path and vice verse
+	Version                  int      `json:"version"`                               // Version number of the role format
+}
+
+func (r *awsRoleEntry) toResponseData() map[string]interface{} {
+	respData := map[string]interface{}{
+		"credential_types": r.CredentialTypes,
+		"policy_arns":      r.PolicyArns,
+		"role_arns":        r.RoleArns,
+		"policy_document":  r.PolicyDocument,
+	}
+	if r.InvalidData != "" {
+		respData["invalid_data"] = r.InvalidData
+	}
+	return respData
```

Comment:
Am I right in presuming it's intentional that `prohibit_flexible_cred_path` and `version` are omitted?

Reply:
Yep. `version` is purely an implementation detail about how the role data is stored in the internal storage, to make upgrading the way roles are stored easier in the future. `prohibit_flexible_cred_path` is also an implementation detail to securely enable backwards compatibility and also unify the `/creds` and `/sts` path. My thinking is, if clients are using the "old" parameters in defining the role, then they'll use the same `/creds` or `/sts` path used before to retrieve credentials. And if they're using the "new" parameters introduced in this PR, then it won't matter which of the two endpoints are used.

Of course, this has the potential to add confusion. I think omitting these is the right balance, but I'd love to hear any suggestions on how it could be improved! :)